crates/walrus-service/src/node/pending_metadata_cache.rs

```diff
@@ -0,0 +1,150 @@
+// Copyright (c) Walrus Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! In-memory cache for metadata awaiting blob registration.
+
+use std::{
+    sync::Arc,
+    time::{Duration, Instant},
+};
+
+use indexmap::IndexMap;
+use tokio::sync::Mutex;
+use walrus_core::{BlobId, metadata::VerifiedBlobMetadataWithId};
+
+use super::metrics::NodeMetricSet;
+
+#[derive(Debug, Clone)]
+struct CachedMetadata {
+    inserted_at: Instant,
+    metadata: Arc<VerifiedBlobMetadataWithId>,
+}
+
+#[derive(Debug)]
+struct PendingMetadataCacheInner {
+    max_entries: usize,
+    ttl: Duration,
+    entries: IndexMap<BlobId, CachedMetadata>,
+}
+
+impl PendingMetadataCacheInner {
+    fn new(max_entries: usize, ttl: Duration) -> Self {
+        Self {
+            max_entries,
+            ttl,
+            entries: IndexMap::new(),
+        }
+    }
+
+    fn evict_expired(&mut self, now: Instant) {
+        if self.ttl.is_zero() {
+            return;
+        }
+        let ttl = self.ttl;
+        self.entries
+            .retain(|_, cached| now.saturating_duration_since(cached.inserted_at) < ttl);
+    }
+
+    fn get(&self, blob_id: &BlobId) -> Option<Arc<VerifiedBlobMetadataWithId>> {
+        self.entries
+            .get(blob_id)
+            .map(|cached| cached.metadata.clone())
+    }
+
+    fn insert(
+        &mut self,
+        blob_id: BlobId,
+        metadata: Arc<VerifiedBlobMetadataWithId>,
+    ) -> Result<(), ()> {
+        let now = Instant::now();
+        self.evict_expired(now);
+        if self.entries.contains_key(&blob_id) {
+            self.entries.insert(
+                blob_id,
+                CachedMetadata {
+                    inserted_at: now,
+                    metadata,
+                },
+            );
+            return Ok(());
+        }
```

Comment:
Nit: The following seems more idiomatic to me:

```suggestion
        if let Entry::Occupied(mut entry) = self.entries.entry(blob_id) {
            entry.insert(CachedMetadata {
                inserted_at: now,
                metadata,
            });
            return Ok(());
        }
```

Potentially, this could also be a match statement where the code below is added to the `Vacant` case.

Reply:
done!